server/auth/azure.js

```diff
@@ -0,0 +1,88 @@
+// @flow
+import passport from "@outlinewiki/koa-passport";
+import jwt from "jsonwebtoken";
+import Router from "koa-router";
+import { capitalize } from "lodash";
+import { Strategy as AzureStrategy } from "passport-azure-ad-oauth2";
+import accountProvisioner from "../commands/accountProvisioner";
+import env from "../env";
+import auth from "../middlewares/authentication";
+import passportMiddleware from "../middlewares/passport";
+import { StateStore } from "../utils/passport";
+
+const router = new Router();
+const providerName = "azure";
+const AZURE_CLIENT_ID = process.env.AZURE_CLIENT_ID;
+const AZURE_CLIENT_SECRET = process.env.AZURE_CLIENT_SECRET;
+
+const scopes = [];
+
+export const config = {
+  name: "Microsoft",
+  enabled: !!AZURE_CLIENT_ID,
+};
+
+if (AZURE_CLIENT_ID) {
+  const strategy = new AzureStrategy(
+    {
+      clientID: AZURE_CLIENT_ID,
+      clientSecret: AZURE_CLIENT_SECRET,
+      callbackURL: `${env.URL}/auth/azure.callback`,
+      useCommonEndpoint: true,
+      passReqToCallback: true,
+      store: new StateStore(),
+      scope: scopes,
+    },
+    async function (req, accessToken, refreshToken, params, _, done) {
+      try {
+        // see docs for what the fields in profile represent here:
+        // https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens
+        const profile = jwt.decode(params.id_token);
+
+        const domain = profile.email.split("@")[1];
+        const subdomain = domain.split(".")[0];
+        const teamName = capitalize(subdomain);
```

Comment:
Ideally we'd grab the organization name with an API request, haven't quite figured that out yet. The accessToken returned from oauth isn't working with Microsoft' graph API

Reply:
@tommoor the callback signature seems to be wrong.
I tried to implement the same strategy, but as you can see on the strategy page (http://www.passportjs.org/packages/passport-azure-oauth2/), there's no `req` object in the callback, the first parameter is the accessToken. So I guess you're trying to query the graph API with the refresh token

Reply:
The startegy callbacks are standardized – `passReqToCallback: true` in the strategy options makes the first parameter always the request.

Reply:
Yes, you're right. The problem actually is that that package use the old Azure AD Graph API, instead of the new Microsoft Graph API

Reply:
Could you point to that? I didn't think it was using either graph api.

I'll probably fork the whole thing, it's also point to an old login url.

Reply:
Sure, here:
https://github.com/AndrewKeig/passport-azure-oauth2/blob/master/lib/index.js#L9
The base URL for oauth requests is `login.windows.net` instead of `graph.microsoft.com`.